src/translator/quality_estimator.cpp

```diff
@@ -0,0 +1,133 @@
+#include <sstream>
+#include "quality_estimator.h"
+
+
+
+namespace marian {
+namespace bergamot {
+
+QualityEstimator::QualityEstimator(const char* file_parameters) {
+  std::string current_str(file_parameters);
+  std::string delimiter = "\n";
+  size_t pos = 0;
+  int line_position = 0;
+  std::string line;
+  while ((pos = current_str.find(delimiter)) != std::string::npos) {
+      line = current_str.substr(0, pos);
+      switch (line_position) {
+        case 0:
+            QualityEstimator::initVector(this->stds, line);
+            break;
+        case 1:
+            QualityEstimator::initVector(this->means, line);
+            break;
+        case 2:
+            QualityEstimator::initVector(this->coefficients, line);
+            break;
+        }
+      current_str.erase(0, pos + delimiter.length());
+      line_position++;
+    }
+    // last line
+   QualityEstimator::initVector(this->intercept, current_str);
+}
+
+void QualityEstimator::initVector(std::vector<float> &emptyVector, std::string line){
+    std::istringstream ss(line);
+  
+    std::string elem;
+    while (ss >> elem) 
+    {
+        emptyVector.push_back(std::stof(elem));
+    }
+}
+
+void QualityEstimator::mapBPEToWords(Response& sentence, Words words){
+    //The result is a vector of maps
+    // each element of vector is a map containing the bpes of a word
+    // each map key is a tuple consisting of the begining and end of the specific subword
+    std::vector<std::map<std::tuple<size_t, size_t>, float>> target_words;
+    for (int sentenceIdx = 0; sentenceIdx < sentence.size(); sentenceIdx++) {
+        auto &quality = sentence.qualityScores[sentenceIdx];
+        size_t wordIdx = 0;
+        bool first = true;
+        std::map<std::tuple<size_t, size_t>, float> map;
+        float previous_p;
+        for (auto &p : quality.word) {
+            if (words[wordIdx].toString()==words[wordIdx].DEFAULT_EOS_ID.toString()){
+                break;
+            }
+            ByteRange subword = sentence.target.wordAsByteRange(sentenceIdx, wordIdx);
+            size_t subword_begin = subword.begin;
+            size_t subword_end = subword.end;
+            char first_word = sentence.target.text[subword_begin];
```

Comment:
There appears to be an assumption that a space can only appear as the first character of a piece.  Appears to be correct, though we should get a documented answer to this.  Also unsure what happens if e.g. there is a tab in the middle of a sentence.

Reply:
I decided to use a regex match instead of a direct string. Please see if you agree: d4345d0

Reply:
Since you're taking the first byte of the string, it will do ASCII spaces.  In which case call `isspace` on the character instead of building a new string from one character then running regex on it.

Reply:
7a52d74